In [1]:
import sys
import yaml
with open('..//secrets.yml', 'r') as file:
    secrets = yaml.safe_load(file)
sys.path.append(secrets['elo_proj_path'])

In [3]:
from player_club_classes import team_elo, Player, Club, Match
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
import plotly
import pickle

from team_colors import team_color_dict
team_colors = pd.DataFrame(team_color_dict).T
team_colors.columns = ['Primary', 'Secondary']
team_colors = team_colors.rename_axis('Team').reset_index()

from plotly.validators.scatter.marker import SymbolValidator
raw_symbols = SymbolValidator().values

simple_symbols = [i for i in raw_symbols if str(i).isalpha()]

with open('../../Rugby_ELO/processed_data/playerbase.pickle', 'rb') as handle:
    playerbase = pickle.load(handle)
with open('../../Rugby_ELO/processed_data/matchlist.pickle', 'rb') as handle:
    matchlist = pickle.load(handle)
with open('../../Rugby_ELO/processed_data/teamlist.pickle', 'rb') as handle:
    teamlist = pickle.load(handle)

In [4]:
match_list = []
for _, match in matchlist.items():
    match_list.append({key:val for key, val in vars(match).items() if key not in ['home_team', 'away_team']})

match_df = pd.DataFrame(match_list)
match_df['week_num'] = match_df['date'].dt.isocalendar().week
match_df['month'] = match_df['date'].dt.month
match_df['year'] = match_df['date'].dt.year

In [62]:
match_list = []
for _, match in matchlist.items():
    match_list.append({key:val for key, val in vars(match).items()})

In [64]:
future_games = [x for x in match_list if 'point_diff' not in x.keys()]

In [91]:
future_game = future_games[0]

home_team = pd.DataFrame(future_game['home_team'][:, [0,1, -1]], columns = ['Number', 'Name', 'elo'])
away_team = pd.DataFrame(future_game['away_team'][:, [0,1, -1]], columns = ['Number', 'Name', 'elo'])
home_team = home_team.sort_values('Number')
away_team = away_team.sort_values('Number')
home_team

Number                    Name         elo
7     1.0           Keita Inagaki   95.693948
8     2.0          Atsushi Sakate  103.016034
9     3.0          Asaeli Ai Valu  109.901054
10    4.0           Warner Dearns   81.632401
13    5.0            Sanaila Waqa   98.572695
12    6.0          Michael Leitch  111.575428
21    7.0              Ben Gunter   74.382303
20    8.0          Jack Cornelsen   100.25964
11    9.0             Naoto Saito   77.317966
6    10.0            Seungsin Lee   78.591693
5    11.0          Siosaia Fifita   74.326031
4    12.0            Shogo Nakano   77.407979
3    13.0             Dylan Riley   81.078527
2    14.0  Gerhard van den Heever   97.958843
0    15.0         Ryohei Yamanaka  105.641678
19   16.0             Shota Horie  100.605033
22   17.0          Yukio Morikawa   79.429971
17   18.0             Yusuke Kizu    81.98869
16   19.0          Takayasu Tsuji          80
15   20.0           Tevita Tatafu  104.821511
14   21.0           Kaito Shigeno   99.664607
18   22.0               Yu Tamura  105.351957
1    23.0             Shane Gates   83.191635

In [96]:
from tabulate import tabulate
home_table = tabulate(home_team, tablefmt="pipe", headers="keys", showindex=False)
away_table = tabulate(away_team, tablefmt="pipe", headers="keys", showindex=False)

In [85]:
f'<table><tr><th>Home Team </th><th>Away Team</th></tr><tr><td>{home_table}</td><td>{away_table}</td></tr> </table>'

|    |   Number | Name                   |      elo |
|---:|---------:|:-----------------------|---------:|
|  7 |        1 | Keita Inagaki          |  95.6939 |
|  8 |        2 | Atsushi Sakate         | 103.016  |
|  9 |        3 | Asaeli Ai Valu         | 109.901  |
| 10 |        4 | Warner Dearns          |  81.6324 |
| 13 |        5 | Sanaila Waqa           |  98.5727 |
| 12 |        6 | Michael Leitch         | 111.575  |
| 21 |        7 | Ben Gunter             |  74.3823 |
| 20 |        8 | Jack Cornelsen         | 100.26   |
| 11 |        9 | Naoto Saito            |  77.318  |
|  6 |       10 | Seungsin Lee           |  78.5917 |
|  5 |       11 | Siosaia Fifita         |  74.326  |
|  4 |       12 | Shogo Nakano           |  77.408  |
|  3 |       13 | Dylan Riley            |  81.0785 |
|  2 |       14 | Gerhard van den Heever |  97.9588 |
|  0 |       15 | Ryohei Yamanaka        | 105.642  |
| 19 |       16 | Shota Horie            | 100.605  |
| 22 |       17 | Yukio Mori

'<table><tr><th>Home Team </th><th>Away Team</th></tr><tr><td>None</td><td>None</td></tr> </table>'

|    |   Number | Name                   |      elo |\n|---:|---------:|:-----------------------|---------:|\n|  7 |        1 | Keita Inagaki          |  95.6939 |\n|  8 |        2 | Atsushi Sakate         | 103.016  |\n|  9 |        3 | Asaeli Ai Valu         | 109.901  |\n| 10 |        4 | Warner Dearns          |  81.6324 |\n| 13 |        5 | Sanaila Waqa           |  98.5727 |\n| 12 |        6 | Michael Leitch         | 111.575  |\n| 21 |        7 | Ben Gunter             |  74.3823 |\n| 20 |        8 | Jack Cornelsen         | 100.26   |\n| 11 |        9 | Naoto Saito            |  77.318  |\n|  6 |       10 | Seungsin Lee           |  78.5917 |\n|  5 |       11 | Siosaia Fifita         |  74.326  |\n|  4 |       12 | Shogo Nakano           |  77.408  |\n|  3 |       13 | Dylan Riley            |  81.0785 |\n|  2 |       14 | Gerhard van den Heever |  97.9588 |\n|  0 |       15 | Ryohei Yamanaka        | 105.642  |\n| 19 |       16 | Shota Horie            | 100.605  |\n| 22 |       17 | Yukio Morikawa         |  79.43   |\n| 17 |       18 | Yusuke Kizu            |  81.9887 |\n| 16 |       19 | Takayasu Tsuji         |  80      |\n| 15 |       20 | Tevita Tatafu          | 104.822  |\n| 14 |       21 | Kaito Shigeno          |  99.6646 |\n| 18 |       22 | Yu Tamura              | 105.352  |\n|  1 |       23 | Shane Gates            |  83.1916 |

<table><tr><th>Home Team </th><th>Away Team</th></tr><tr><td>
|   Number | Name                   |      elo |\n|---:|---------:|:-----------------------|---------:|\n|  7 |        1 | Keita Inagaki          |  95.6939 |\n|  8 |        2 | Atsushi Sakate         | 103.016  |\n|  9 |        3 | Asaeli Ai Valu         | 109.901  |\n| 10 |        4 | Warner Dearns          |  81.6324 |\n| 13 |        5 | Sanaila Waqa           |  98.5727 |\n| 12 |        6 | Michael Leitch         | 111.575  |\n| 21 |        7 | Ben Gunter             |  74.3823 |\n| 20 |        8 | Jack Cornelsen         | 100.26   |\n| 11 |        9 | Naoto Saito            |  77.318  |\n|  6 |       10 | Seungsin Lee           |  78.5917 |\n|  5 |       11 | Siosaia Fifita         |  74.326  |\n|  4 |       12 | Shogo Nakano           |  77.408  |\n|  3 |       13 | Dylan Riley            |  81.0785 |\n|  2 |       14 | Gerhard van den Heever |  97.9588 |\n|  0 |       15 | Ryohei Yamanaka        | 105.642  |\n| 19 |       16 | Shota Horie            | 100.605  |\n| 22 |       17 | Yukio Morikawa         |  79.43   |\n| 17 |       18 | Yusuke Kizu            |  81.9887 |\n| 16 |       19 | Takayasu Tsuji         |  80      |\n| 15 |       20 | Tevita Tatafu          | 104.822  |\n| 14 |       21 | Kaito Shigeno          |  99.6646 |\n| 18 |       22 | Yu Tamura              | 105.352  |\n|  1 |       23 | Shane Gates            |  83.1916 |
</td><td>
|    |   Number | Name               |      elo |\n|---:|---------:|:-------------------|---------:|\n|  5 |        1 | Jean-Baptiste Gros | 114.835  |\n|  6 |        2 | Peato Mauvaka      | 108.294  |\n|  7 |        3 | Demba Bamba        | 106.964  |\n|  8 |        4 | Thibaud Flament    |  91.3623 |\n|  0 |        5 | Thomas Jolmes      |  93.8563 |\n|  9 |        6 | Dylan Cretin       |  89.5152 |\n| 20 |        7 | Charles Ollivon    | 124.007  |\n| 11 |        8 | Yoan Tanga         | 103.367  |\n|  4 |        9 | Maxime Lucu        |  88.4409 |\n|  3 |       10 | Matthieu Jalibert  | 110.789  |\n|  2 |       11 | Matthis Lebel      |  88.1638 |\n|  1 |       12 | Yoram Moefana      |  83.9963 |\n| 19 |       13 | Virimi Vakatawa    | 117.642  |\n| 21 |       14 | Damian Penaud      | 106.633  |\n| 22 |       15 | Max Spring         |  83.4149 |\n| 12 |       16 | Pierre Bourgarit   | 104.391  |\n| 13 |       17 | Dany Priso         | 104.041  |\n| 14 |       18 | Sipili Falatea     | 102.637  |\n| 15 |       19 | Thomas Lavault     |  86.7407 |\n| 16 |       20 | Ibrahim Diallo     |  76.6319 |\n| 17 |       21 | Sekou Macalou      | 114.735  |\n| 18 |       22 | Baptiste Couilloud | 118.945  |\n| 10 |       23 | Antoine Hastoy     |  94.7028 |
</td></tr> </table>

<table><tr><th>Table 1 Heading 1 </th><th>Table 1 Heading 2</th></tr><tr><td>

|Table 1| Middle | Table 2|
|--|--|--|
|a| not b|and c |

</td><td>

|b|1|2|3| 
|--|--|--|--|
|a|s|d|f|

</td></tr> </table>

In [39]:
future_game["date"].date().strftime("%Y-%m-%d")

'2022-07-08'

In [30]:
f'{future_game["away_team_name"]} ({round(future_game["lineup_away_elo"], 2)}) at {future_game["home_team_name"]} ({round(future_game["lineup_home_elo"], 2)})'
favorite = future_game["home_team_name"] if future_game["lineup_spread"] > 0 else future_game["away_team_name"]
f'Prediction: {favorite} by {round(abs(future_game["lineup_spread"]), 1)}'

'Prediction: Wales by 3.5'

0.47967167180130527

In [6]:
month = match_df[(match_df.year == 2022) & (match_df.month == 7)]
closest_games = month.sort_values('lineup_spread',  key=abs).head()

sorted_pred_accuracy = month.assign(tmp=month["point_diff"] - month["lineup_spread"]).sort_values(by="tmp", key=abs)#.drop(columns="tmp")
best_preds = sorted_pred_accuracy.head()
worst_preds = sorted_pred_accuracy.tail()

biggest_upsets = month[(month.lineup_spread * month.point_diff) < 0].assign(tmp=month["point_diff"] - month["lineup_spread"]).sort_values(by="tmp", key=abs, ascending=False).drop(columns="tmp").head()
biggest_covers = month[((month.lineup_spread * month.point_diff) > 0)].assign(tmp=month["point_diff"] - month["lineup_spread"]).sort_values(by="tmp", key=abs, ascending=False)#.drop(columns="tmp").head()

In [8]:
player_elo_list = []
for player_name, player in playerbase.items():
    player_elo = pd.DataFrame(player.elo_list, columns = [
        'Number', 'Full_Name', 'Team', 'Player', 'Position', 'Tries',
        'Try Assists', 'Conversion Goals', 'Penalty Goals',
        'Drop Goals Converted', 'Points', 'Passes', 'Runs', 'Meters Run',
        'Clean Breaks', 'Defenders Beaten', 'Offload', 'Turnovers Conceded',
        'Tackles', 'Missed Tackles', 'Lineouts Won', 'Penalties Conceded',
        'Yellow Cards', 'Red Cards', 'espn_id_num', 'Competition', 'Date',
        'Home Team', 'Home Score', 'Away Team', 'Away Score', 'Minutes',
        'Position_Number', 'gameid', 'Unicode_ID', 'start_elo', 'end_elo'
       ])
    player_elo['Full Name'] = player_name[0]
    player_elo['Unicode_ID'] = player_name[1]
    player_elo_list.append(player_elo)

player_elo = pd.concat(player_elo_list).reset_index(drop=True)
player_elo = pd.merge(player_elo, team_colors, on = 'Team', how = 'left')
player_elo['elo_change'] = player_elo.end_elo - player_elo.start_elo
player_elo.Date = pd.to_datetime(player_elo.Date)

player_elo['week_num'] = player_elo['Date'].dt.isocalendar().week
player_elo['month'] = player_elo['Date'].dt.month
player_elo['year'] = player_elo['Date'].dt.year

In [9]:
player_elo.tail()

Number               Full_Name     Team          Player Position Tries  \
503976    5.0  Konstantine Mikautadze  Georgia    K Mikautadze        L   0.0   
503977   14.0          Aka Tabutsadze  Georgia    A Tabutsadze        W   0.0   
503978   20.0        Luka Ivanishvili  Georgia   L Ivanishvili      NaN   0.0   
503979   11.0         Alexander Todua  Georgia         A Todua        W   0.0   
503980   17.0       Nika Khatiashvili  Georgia  N Khatiashvili      NaN   0.0   

       Try Assists Conversion Goals Penalty Goals Drop Goals Converted  ...  \
503976         0.0              0.0           0.0                  0.0  ...   
503977         0.0              0.0           0.0                  0.0  ...   
503978         0.0              0.0           0.0                  0.0  ...   
503979         0.0              0.0           0.0                  0.0  ...   
503980         0.0              0.0           0.0                  0.0  ...   

                                               Unicode_ID start_elo  \
503976  1101119510111511211095105100327511111011511697...        80   
503977  1101119510111511211095105100326510797328497981...        80   
503978  1101119510111511211095105100327611710797327311...        80   
503979  1101119510111511211095105100326510810112097110...        80   
503980  1101119510111511211095105100327810510797327510...        80   

          end_elo               Full Name  Primary Secondary elo_change  \
503976  81.113136  Konstantine Mikautadze  #551828   #231f20   1.113136   
503977  81.113136          Aka Tabutsadze  #551828   #231f20   1.113136   
503978       80.0        Luka Ivanishvili  #551828   #231f20        0.0   
503979  81.113136         Alexander Todua  #551828   #231f20   1.113136   
503980       80.0       Nika Khatiashvili  #551828   #231f20        0.0   

       week_num month  year  
503976       26     7  2022  
503977       26     7  2022  
503978       26     7  2022  
503979       26     7  2022  
503980       26     7  2022  

[5 rows x 44 columns]

In [7]:
# CHANGE
player_month = player_elo[(player_elo.year == 2022) & (player_elo.month == 5)]
oneplayer = player_month[(player_month.Unicode_ID == '5155544848') & (player_month['Full Name'] == 'Caleb Delany')].sort_values('Date')
oneplayer.iloc[-1].end_elo - oneplayer.iloc[0].start_elo

-2.554117441069806

In [11]:
def elo_change(df):
    return df.iloc[-1].end_elo, df.iloc[0].start_elo, df.iloc[-1].end_elo - df.iloc[0].start_elo

elo_changes = player_month.sort_values('Date').groupby(['Unicode_ID', 'Full Name', 'Team']).apply(elo_change).reset_index()
elo_changes[['end_elo', 'start_elo', 'elo_change']] = pd.DataFrame(elo_changes[0].tolist(), index=elo_changes.index)
elo_changes.sort_values('elo_change')

Unicode_ID          Full Name              Team  \
576   5051515156    James Blackwell        Hurricanes   
826   5055485550      Salesi Rayasi        Hurricanes   
1070  5149525654       Quinn Tupaea            Chiefs   
267   5048485452        Ardie Savea        Hurricanes   
589   5051515748         Josh Ioane            Chiefs   
...          ...                ...               ...   
406   5049535052     Mattia Bellini  Benetton Treviso   
745   5053505254        Rhyno Smith  Benetton Treviso   
244   4957555455  Juan Ignacio Brex  Benetton Treviso   
408   5049535155     Federico Ruzza  Benetton Treviso   
1333  5156565456      Vinaya Habosi       Fijian Drua   

                                                      0     end_elo  \
576   (100.88391371265058, 106.42990302317858, -5.54...  100.883914   
826   (110.70188334521939, 115.67822631062886, -4.97...  110.701883   
1070  (113.26707089934938, 118.08496318931884, -4.81...  113.267071   
267   (118.33886733853684, 122.95562106235826, -4.61...  118.338867   
589   (98.22976447566921, 102.77928260508381, -4.549...   98.229764   
...                                                 ...         ...   
406   (96.79422627952788, 92.30762615721996, 4.48660...   96.794226   
745   (99.09602275738854, 94.60942263508062, 4.48660...   99.096023   
244   (98.82188475913395, 94.33528463682603, 4.48660...   98.821885   
408   (102.4315904715746, 97.94499034926667, 4.48660...  102.431590   
1333  (98.67315216377675, 94.10948489797238, 4.56366...   98.673152   

       start_elo  elo_change  
576   106.429903   -5.545989  
826   115.678226   -4.976343  
1070  118.084963   -4.817892  
267   122.955621   -4.616754  
589   102.779283   -4.549518  
...          ...         ...  
406    92.307626    4.486600  
745    94.609423    4.486600  
244    94.335285    4.486600  
408    97.944990    4.486600  
1333   94.109485    4.563667  

[1432 rows x 7 columns]